core/src/main/scala/kafka/log/CleanerCache.scala

```diff
@@ -0,0 +1,289 @@
+  def latestOffset: Long
+
+  /**
+   * @return True if the passed record has a larger version (or offset, if no version exists)
+   *         than the currently cached record, or if no cached record exists at all
+   */
+  def greater(record: Record): Boolean
+}
+
+/**
+ * An hash table used for deduplicating the log. This hash table uses a cryptographicly secure hash of the key as a proxy for the key
+ * for comparisons and to save space on object overhead. Collisions are resolved by probing. This hash table does not support deletes.
+ *
+ * @param memory        The amount of memory this map can use
+ * @param hashAlgorithm The hash algorithm instance to use: MD2, MD5, SHA-1, SHA-256, SHA-384, SHA-512
+ * @param strategy      The compaction strategy for this cleaner to adopt
+ */
+@nonthreadsafe
+class SkimpyCleanerCache(val memory: Int, val hashAlgorithm: String = "MD5", val strategy: String = Defaults.CompactionStrategy) extends CleanerCache {
+  private val bytes = ByteBuffer.allocate(memory)
+
+  /* the hash algorithm instance to use, default is MD5 */
+  private val digest = MessageDigest.getInstance(hashAlgorithm)
+
+  /* the number of bytes for this hash algorithm */
+  private val hashSize = digest.getDigestLength
+
+  /* create some hash buffers to avoid reallocating each time */
+  private val hash1 = new Array[Byte](hashSize)
+  private val hash2 = new Array[Byte](hashSize)
+
+  /* number of entries put into the map */
+  private var entries = 0
+
+  /* number of lookups on the map */
+  private var lookups = 0L
+
+  /* the number of probes for all lookups */
+  private var probes = 0L
+
+  /* the latest offset written into the map */
+  private var lastOffset = -1L
+
+  /**
+   * The number of bytes of space each entry uses.
+   * This evaluates to the number of bytes in the hash plus 8 bytes for the offset
+   * and, if applicable, another 8 bytes for the version header.
+   */
+  val bytesPerEntry: Int = hashSize + 8 + (if (enhancedCache) 8 else 0)
+
+  val slots: Int = memory / bytesPerEntry
+
+  override def putIfGreater(record: Record): Boolean = {
+    if (!record.hasKey || !greater(record)) {
+      return false
+    }
+    val recordKey = record.key
+    val recordOffset = record.offset
+    val recordVersion = extractVersion(record)
+    require(entries < slots, "Attempt to add a new entry to a full offset map.")
+    lookups += 1
+    hashInto(recordKey, hash1)
+    // probe until we find the first empty slot
+    var attempt = 0
+    var pos = positionOf(hash1, attempt)
+    while (!isEmpty(pos)) {
+      bytes.position(pos)
+      bytes.get(hash2)
+      if (util.Arrays.equals(hash1, hash2)) {
+        // we found an existing entry, overwrite it and return (size does not change)
+        bytes.putLong(recordOffset)
+        if (enhancedCache)
+          bytes.putLong(recordVersion)
+        updateLatestOffset(recordOffset)
+        return true
+      }
+      attempt += 1
+      pos = positionOf(hash1, attempt)
```

Comment:
When would the loop exit if cache is full ?

Reply:
This loop is broken when no entries are found at the current 'pos'.
Unit tests fail if this is not reached (i.e., if you have an infinite loop or a no-op loop).